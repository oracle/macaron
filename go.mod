// Copyright (c) 2022 - 2022, Oracle and/or its affiliates. All rights reserved.
// Licensed under the Universal Permissive License v 1.0 as shown at https://oss.oracle.com/licenses/upl/.

module github.com/oracle-samples/macaron

go 1.18

require (
	github.com/rhysd/actionlint v1.6.22
	mvdan.cc/sh/v3 v3.6.0
)

require (
	github.com/fatih/color v1.13.0 // indirect
	github.com/mattn/go-colorable v0.1.13 // indirect
	github.com/mattn/go-isatty v0.0.16 // indirect
	github.com/mattn/go-runewidth v0.0.14 // indirect
	github.com/rivo/uniseg v0.4.3 // indirect
	github.com/robfig/cron v1.2.0 // indirect
	golang.org/x/sync v0.1.0 // indirect
	golang.org/x/sys v0.3.0 // indirect
	gopkg.in/check.v1 v1.0.0-20180628173108-788fd7840127 // indirect
	gopkg.in/yaml.v3 v3.0.1 // indirect
)
