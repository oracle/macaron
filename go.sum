github.com/fatih/color v1.13.0 h1:8LOYc1KYPPmyKMuN8QV2DNRWNbLo6LZ0iLs8+mlH53w=
github.com/fatih/color v1.13.0/go.mod h1:kLAiJbzzSOZDVNGyDpeOxJ47H46qBXwg5ILebYFFOfk=
github.com/frankban/quicktest v1.14.4 h1:g2rn0vABPOOXmZUj+vbmUp0lPoXEMuhTpIluN0XL9UY=
github.com/google/go-cmp v0.5.9 h1:O2Tfq5qg4qc4AmwVlvv0oLiVAGB7enBSJ2x2DqQFi38=
github.com/kr/pretty v0.3.1 h1:flRD4NNwYAUpkphVc1HcthR4KEIFJ65n8Mw5qdRn3LE=
github.com/kr/text v0.2.0 h1:5Nx0Ya0ZqY2ygV366QzturHI13Jq95ApcVaJBhpS+AY=
github.com/mattn/go-colorable v0.1.9/go.mod h1:u6P/XSegPjTcexA+o6vUJrdnUu04hMope9wVRipJSqc=
github.com/mattn/go-colorable v0.1.13 h1:fFA4WZxdEF4tXPZVKMLwD8oUnCTTo08duU7wxecdEvA=
github.com/mattn/go-colorable v0.1.13/go.mod h1:7S9/ev0klgBDR4GtXTXX8a3vIGJpMovkB8vQcUbaXHg=
github.com/mattn/go-isatty v0.0.12/go.mod h1:cbi8OIDigv2wuxKPP5vlRcQ1OAZbq2CE4Kysco4FUpU=
github.com/mattn/go-isatty v0.0.14/go.mod h1:7GGIvUiUoEMVVmxf/4nioHXj79iQHKdU27kJ6hsGG94=
github.com/mattn/go-isatty v0.0.16 h1:bq3VjFmv/sOjHtdEhmkEV4x1AJtvUvOJ2PFAZ5+peKQ=
github.com/mattn/go-isatty v0.0.16/go.mod h1:kYGgaQfpe5nmfYZH+SKPsOc2e4SrIfOl2e/yFXSvRLM=
github.com/mattn/go-runewidth v0.0.14 h1:+xnbZSEeDbOIg5/mE6JF0w6n9duR1l3/WmbinWVwUuU=
github.com/mattn/go-runewidth v0.0.14/go.mod h1:Jdepj2loyihRzMpdS35Xk/zdY8IAYHsh153qUoGf23w=
github.com/rhysd/actionlint v1.6.22 h1:cAEf2PGNwJXhdcTVF2xS/0ORqWS+ueUHwjQYsqFsGSk=
github.com/rhysd/actionlint v1.6.22/go.mod h1:gIKOdxtV40mBOcD0ZR8EBa8NqjEXToAZioroS3oedMg=
github.com/rivo/uniseg v0.2.0/go.mod h1:J6wj4VEh+S6ZtnVlnTBMWIodfgj8LQOQFoIToxlJtxc=
github.com/rivo/uniseg v0.4.3 h1:utMvzDsuh3suAEnhH0RdHmoPbU648o6CvXxTx4SBMOw=
github.com/rivo/uniseg v0.4.3/go.mod h1:FN3SvrM+Zdj16jyLfmOkMNblXMcoc8DfTHruCPUcx88=
github.com/robfig/cron v1.2.0 h1:ZjScXvvxeQ63Dbyxy76Fj3AT3Ut0aKsyd2/tl3DTMuQ=
github.com/robfig/cron v1.2.0/go.mod h1:JGuDeoQd7Z6yL4zQhZ3OPEVHB7fL6Ka6skscFHfmt2k=
github.com/rogpeppe/go-internal v1.9.0 h1:73kH8U+JUqXU8lRuOHeVHaa/SZPifC7BkcraZVejAe8=
golang.org/x/sync v0.1.0 h1:wsuoTGHzEhffawBOhz5CYhcrV4IdKZbEyZjBMuTp12o=
golang.org/x/sync v0.1.0/go.mod h1:RxMgew5VJxzue5/jJTE5uejpjVlOe/izrB70Jof72aM=
golang.org/x/sys v0.0.0-20200116001909-b77594299b42/go.mod h1:h1NjWce9XRLGQEsW7wpKNCjG9DtNlClVuFLEZdDNbEs=
golang.org/x/sys v0.0.0-20200223170610-d5e6a3e2c0ae/go.mod h1:h1NjWce9XRLGQEsW7wpKNCjG9DtNlClVuFLEZdDNbEs=
golang.org/x/sys v0.0.0-20210630005230-0f9fa26af87c/go.mod h1:oPkhp1MJrh7nUepCBck5+mAzfO9JrbApNNgaTdGDITg=
golang.org/x/sys v0.0.0-20220811171246-fbc7d0a398ab/go.mod h1:oPkhp1MJrh7nUepCBck5+mAzfO9JrbApNNgaTdGDITg=
golang.org/x/sys v0.3.0 h1:w8ZOecv6NaNa/zC8944JTU3vz4u6Lagfk4RPQxv92NQ=
golang.org/x/sys v0.3.0/go.mod h1:oPkhp1MJrh7nUepCBck5+mAzfO9JrbApNNgaTdGDITg=
gopkg.in/check.v1 v0.0.0-20161208181325-20d25e280405/go.mod h1:Co6ibVJAznAaIkqp8huTwlJQCZ016jof/cbN4VW5Yz0=
gopkg.in/check.v1 v1.0.0-20180628173108-788fd7840127 h1:qIbj1fsPNlZgppZ+VLlY7N33q108Sa+fhmuc+sWQYwY=
gopkg.in/check.v1 v1.0.0-20180628173108-788fd7840127/go.mod h1:Co6ibVJAznAaIkqp8huTwlJQCZ016jof/cbN4VW5Yz0=
gopkg.in/yaml.v3 v3.0.1 h1:fxVm/GzAzEWqLHuvctI91KS9hhNmmWOoWu0XTYJS7CA=
gopkg.in/yaml.v3 v3.0.1/go.mod h1:K4uyk7z7BCEPqu6E+C64Yfv1cQ7kz7rIZviUmN+EgEM=
mvdan.cc/sh/v3 v3.6.0 h1:gtva4EXJ0dFNvl5bHjcUEvws+KRcDslT8VKheTYkbGU=
mvdan.cc/sh/v3 v3.6.0/go.mod h1:U4mhtBLZ32iWhif5/lD+ygy1zrgaQhUu+XFy7C8+TTA=
