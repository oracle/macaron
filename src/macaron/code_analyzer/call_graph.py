# Copyright (c) 2022 - 2024, Oracle and/or its affiliates. All rights reserved.
# Licensed under the Universal Permissive License v 1.0 as shown at https://oss.oracle.com/licenses/upl/.

"""This module contains classes to generate build call graphs for the target repository."""

from collections import deque
from collections.abc import Iterable
from typing import Any, Generic, TypeVar

Node = TypeVar("Node", bound="BaseNode")
# The documentation below for `TypeVar` is commented out due to a breaking
# change in Sphinx version (^=6.1.0).
# Reported at: https://github.com/oracle/macaron/issues/58.
# """This binds type ``Node`` to ``BaseNode`` and any of its subclasses.

# Therefore, any node of type ``Node`` that is stored in the call graph
# container will be a subtype of ``BaseNode``.
# """


class BaseNode(Generic[Node]):
    """This is the generic class for call graph nodes."""

    def __init__(self, caller: Node | None = None, node_id: str | None = None) -> None:
        """Initialize instance.

        Parameter
        ---------
        caller: Node | None
            The caller node.
        node_id: str | None
            The unique identifier of a node in the callgraph.
        """
        self.callee: list[Node] = []
        self.caller: Node | None = caller
        # Each node can have a model that summarizes certain properties for static analysis.
        # By default this model is set to None.
        self.model: Any = None
        self.node_id = node_id

    def add_callee(self, node: Node) -> None:
        """Add a callee to the current node.

        Parameters
        ----------
        node : Node
            The callee node.
        """
        self.callee.append(node)

    def has_callee(self) -> bool:
        """Check if the current node has callees.

        Returns
        -------
        bool
            Return False if there are no callees, otherwise True.
        """
        return bool(self.callee)


class CallGraph(Generic[Node]):
    """This is the generic class for creating a call graph."""

    def __init__(self, root: Node, repo_path: str) -> None:
        """Initialize instance.

        Parameters
        ----------
        root : Node
            The root call graph node.
        repo_path : str
            The path to the repo.
        """
        self.root = root
        self.repo_path = repo_path

    def get_root(self) -> Node:
        """Get the root node in the call graph.

        Returns
        -------
        Node
            The root node.
        """
        return self.root

    def bfs(self) -> Iterable[Node]:
        """Traverse the call graph in breadth first search order.

        Yields
        ------
        Node
            The traversed nodes.
        """
        queue: deque[Node] = deque()
        queue.extend(self.root.callee)
        visited = []
        while queue:
            node = queue.popleft()
            if node not in visited:
                queue.extend(node.callee)
                visited.append(node)
                yield node
