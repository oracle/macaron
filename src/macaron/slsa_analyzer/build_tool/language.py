# Copyright (c) 2024 - 2024, Oracle and/or its affiliates. All rights reserved.
# Licensed under the Universal Permissive License v 1.0 as shown at https://oss.oracle.com/licenses/upl/.

"""This module contains abstractions for build languages."""

from enum import Enum
from typing import Protocol, runtime_checkable


class BuildLanguage(str, Enum):
    """The supported build languages."""

    JAVA = "java"
    PYTHON = "python"
    GO = "go"
    JAVASCRIPT = "javascript"
    DOCKER = "docker"


@runtime_checkable
class Language(Protocol):
    """Interface of a language."""

    @property
    def lang_name(self) -> str:
        """Get the name of the language."""

    @property
    def lang_versions(self) -> list[str] | None:
        """Get the possible versions of the language."""

    @property
    def lang_distributions(self) -> list[str] | None:
        """Get the possible distributions of the language."""

    @property
    def lang_url(self) -> str | None:
        """Get the URL that provides information about the language distributions and versions."""
