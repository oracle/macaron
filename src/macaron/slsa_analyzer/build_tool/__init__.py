# Copyright (c) 2022 - 2022, Oracle and/or its affiliates. All rights reserved.
# Licensed under the Universal Permissive License v 1.0 as shown at https://oss.oracle.com/licenses/upl/.

"""The build_tool package contains the supported build tools for Macaron."""

from .base_build_tool import BaseBuildTool
from .gradle import Gradle
from .maven import Maven

# The list of supported build tools. The order of the list determine the order
# in which each build tool is checked against the target repository.
BUILD_TOOLS: list[BaseBuildTool] = [Gradle(), Maven()]
