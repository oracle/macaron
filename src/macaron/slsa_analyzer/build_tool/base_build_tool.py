# Copyright (c) 2022 - 2024, Oracle and/or its affiliates. All rights reserved.
# Licensed under the Universal Permissive License v 1.0 as shown at https://oss.oracle.com/licenses/upl/.

"""This module contains the BaseBuildTool class to be inherited by other specific Build Tools."""

import glob
import itertools
import json
import logging
import os
from abc import ABC, abstractmethod
from collections.abc import Iterable
from dataclasses import dataclass
from pathlib import Path
from typing import TypedDict

from macaron.code_analyzer.call_graph import BaseNode
from macaron.dependency_analyzer import DependencyAnalyzer
from macaron.slsa_analyzer.build_tool.language import BuildLanguage
from macaron.slsa_analyzer.checks.check_result import Confidence, Evidence, EvidenceWeightMap

logger: logging.Logger = logging.getLogger(__name__)


class BuildToolCommand(TypedDict):
    """This class is an abstraction for build tool commands storing useful contextual data for analysis."""

    #: The parsed build tool command. This command can be any bash command whose program name is the build tool.
    command: list[str]

    #: The name of the language to build the artifact.
    language: str

    #: The list of possible language version numbers.
    language_versions: list[str] | None

    #: The list of possible language distributions.
    language_distributions: list[str] | None

    #: The URL providing information about the language distributions and versions.
    language_url: str | None

    #: The relative path to the root CI file that ultimately triggers the command.
    ci_path: str

    #: The CI step object that calls the command.
    step_node: BaseNode

    #: The list of name of reachable variables that contain secrets."""
    reachable_secrets: list[str]

    #: The name of CI events that trigger the workflow running the build command.
    events: list[str] | None


def file_exists(path: str, file_name: str) -> bool:
    """Return True if a file exists in a directory.

    This method searches in the directory recursively.

    Parameters
    ----------
    path : str
        The path to search for the file.
    file_name : str
        The name of the file to search.

    Returns
    -------
    bool
        True if file_name exists else False.
    """
    pattern = os.path.join(path, "**", file_name)
    files_detected = glob.iglob(pattern, recursive=True)
    try:
        next(files_detected)
        return True
    except StopIteration:
        return False


@dataclass
class RuntimeOptions:
    """The class for build tool runtime configurations read from `defaults.ini`.

    Note that Macaron uses the options in this class to "run" a build tool.
    """

    #: The timeout used for running the build tool commands.
    build_timeout: float = 600


class BaseBuildTool(ABC):
    """This abstract class is used to implement Build Tools."""

    def __init__(self, name: str, language: BuildLanguage) -> None:
        """Initialize instance.

        Parameters
        ----------
        name : str
            The name of this build tool.
        language: BuildLanguage
            The name of the language used by the programs built by the build tool.
        """
        self.name = name
        self.language = language
        self.entry_conf: list[str] = []
        self.build_configs: list[str] = []
        self.package_lock: list[str] = []
        self.builder: list[str] = []
        self.packager: list[str] = []
        self.publisher: list[str] = []
        self.interpreter: list[str] = []
        self.interpreter_flag: list[str] = []
        self.build_arg: list[str] = []
        self.deploy_arg: list[str] = []
        self.ci_build_kws: dict[str, list[str]] = {
            "github_actions": [],
            "travis_ci": [],
            "circle_ci": [],
            "gitlab_ci": [],
            "jenkins": [],
        }
        self.ci_deploy_kws: dict[str, list[str]] = {
            "github_actions": [],
            "travis_ci": [],
            "circle_ci": [],
            "gitlab_ci": [],
            "jenkins": [],
        }
        self.build_log: list[str] = []
        self.wrapper_files: list[str] = []
        self.runtime_options = RuntimeOptions()

    def __str__(self) -> str:
        return self.name

    @abstractmethod
    def is_detected(self, repo_path: str) -> bool:
        """Return True if this build tool is used in the target repo.

        Parameters
        ----------
        repo_path : str
            The path to the target repo.

        Returns
        -------
        bool
            True if this build tool is detected, else False.
        """

    @abstractmethod
    def prepare_config_files(self, wrapper_path: str, build_dir: str) -> bool:
        """Prepare the necessary wrapper files for running the build.

        This method will return False if there is any errors happened during operation.

        Parameters
        ----------
        wrapper_path : str
            The path where all necessary wrapper files are located.
        build_dir : str
            The path of the build dir. This is where all files are copied to.

        Returns
        -------
        bool
            True if succeed else False.
        """

    @abstractmethod
    def load_defaults(self) -> None:
        """Load the default values from defaults.ini."""

    @abstractmethod
    def get_dep_analyzer(self, repo_path: str) -> DependencyAnalyzer:
        """Create a DependencyAnalyzer for the build tool.

        Parameters
        ----------
        repo_path: str
            The path to the target repo.

        Returns
        -------
        DependencyAnalyzer
            The DependencyAnalyzer object.
        """

    def get_build_dirs(self, repo_path: str) -> Iterable[Path]:
        """Find directories in the repository that have their own build scripts.

        This is especially important for applications that consist of multiple services.

        Parameters
        ----------
        repo_path: str
            The path to the target repo.

        Yields
        ------
        Path
            The relative paths from the repo path that contain build scripts.
        """
        config_paths: set[str] = set()
        for build_cfg in self.build_configs:
            config_paths.update(
                path
                for path in glob.glob(os.path.join(repo_path, "**", build_cfg), recursive=True)
                if self.is_detected(str(Path(path).parent))
            )

        list_iter = iter(sorted(config_paths, key=lambda x: (str(Path(x).parent), len(Path(x).parts))))
        try:
            cfg_path = next(list_iter)
            yield Path(cfg_path).parent.relative_to(repo_path)
            while next_item := next(list_iter):
                if next_item.startswith(str(Path(cfg_path).parent)):
                    continue
                cfg_path = next_item
                yield Path(next_item).parent.relative_to(repo_path)

        except StopIteration:
            pass

    def serialize_to_json(self, cmd: list[str]) -> str:
        """Convert a list of values to a json-encoded string so that it is easily parsable by later consumers.

        Parameters
        ----------
        cmd: list[str]
            List of command-line arguments.

        Returns
        -------
        str
            The list of command-line arguments as a json-encoded string.
        """
        return json.dumps(cmd)

    def is_build_command(self, cmd: list[str]) -> bool:
        """
        Determine if the command is a build tool command.

        Parameters
        ----------
        cmd: list[str]
            List of command-line arguments.

        Returns
        -------
        bool
            True if the command is a build tool command.
        """
        # Check for empty or invalid commands.
        if not cmd or not cmd[0]:
            return False
        # The first argument in a bash command is the program name.
        # So first check that the program name is a supported build tool name.
        # We need to handle cases where the first argument is a path to the program.
        cmd_program_name = os.path.basename(cmd[0])
        if not cmd_program_name:
            logger.debug("Found invalid program name %s.", cmd[0])
            return False

        build_tools = set(itertools.chain(self.builder, self.packager, self.publisher, self.interpreter))
        if any(tool for tool in build_tools if tool == cmd_program_name):
            return True

        return False

    def match_cmd_args(self, cmd: list[str], tools: list[str], args: list[str]) -> bool:
        """
        Check if the build command matches any of the tools and the command-line arguments.

        If build command's first element, which is the program name matches any of the `tools` names and any of its arguments
        match any of the arguments in `args`, this function returns True.

        Parameters
        ----------
        cmd: list[str]
            The command-line arguments.
        tools: list[str]
            The name of tools that will be matched with the program name in the bash command.
        args: list[str]
            The lit of arguments that should match with the bash command.

        Returns
        -------
        bool
            True if the provided command matches the tool and arguments.
        """
        cmd_program_name = os.path.basename(cmd[0])

        if cmd_program_name in tools:
            # Check the arguments in the bash command.
            # If there are no args expected for this build tool, accept the command.
            if not args:
                logger.debug("No build arguments required. Accept the %s command.", self.serialize_to_json(cmd))
                return True

            for word in cmd[1:]:
                # TODO: allow plugin versions in arguments, e.g., maven-plugin:1.6.8:deploy.
                if word in args:
                    logger.debug("Found the command %s.", self.serialize_to_json(cmd))
                    return True

        return False

    def infer_confidence_deploy_command(self, cmd: BuildToolCommand) -> Confidence:
        """
        Infer the confidence level for the deploy command.

        Parameters
        ----------
        cmd: BuildToolCommand
            The build tool command object.

        Returns
        -------
        Confidence
            The confidence level for the deploy command.
        """
        # Apply heuristics and assign weights and scores for the discovered evidence.
        # TODO: infer the scores based on existing data using probabilistic inference.
        # Initialize the map.
        evidence_weight_map = EvidenceWeightMap(
            [
                Evidence(name="reachable_secrets", found=False, weight=1),
                Evidence(name="ci_workflow_name", found=False, weight=2),
                Evidence(name="release_event", found=False, weight=2),
            ]
        )

        # Check if secrets are present in the caller job.
        if cmd["reachable_secrets"]:
            evidence_weight_map.update_result(name="reachable_secrets", found=True)

        # Check workflow names.
        deploy_keywords = ["release", "deploy", "publish"]
        for kw in deploy_keywords:
            if kw in os.path.basename(cmd["ci_path"]).lower():
                evidence_weight_map.update_result(name="ci_workflow_name", found=True)
                break

        # Check if the CI workflow is triggered by a release event.
        if cmd["events"] and "release" in cmd["events"]:
            evidence_weight_map.update_result(name="release_event", found=True)

        return Confidence.normalize(evidence_weight_map=evidence_weight_map)

    def is_deploy_command(
        self, cmd: BuildToolCommand, excluded_configs: list[str] | None = None
    ) -> tuple[bool, Confidence]:
        """
        Determine if the command is a deploy command.

        A deploy command usually performs multiple tasks, such as compilation, packaging, and publishing the artifact.
        This function filters the build tool commands that are called from the configuration files provided as input.

        Parameters
        ----------
        cmd: BuildToolCommand
            The build tool command object.
        excluded_configs: list[str] | None
            Build tool commands that are called from these configuration files are excluded.

        Returns
        -------
        tuple[bool, Confidence]
            Return True along with the inferred confidence level if the command is a deploy tool command.
        """
        # Check the language.
        if cmd["language"] is not self.language:
            return False, Confidence.HIGH
        # Some projects use a publisher tool and some use the build tool with deploy arguments.
        deploy_tool = self.publisher if self.publisher else self.builder

        if not self.match_cmd_args(cmd=cmd["command"], tools=deploy_tool, args=self.deploy_arg):
            return False, Confidence.HIGH

        # Check if the CI workflow is a configuration for a known tool.
        if excluded_configs and os.path.basename(cmd["ci_path"]) in excluded_configs:
            return False, Confidence.HIGH

        return True, self.infer_confidence_deploy_command(cmd=cmd)

    def is_package_command(
        self, cmd: BuildToolCommand, excluded_configs: list[str] | None = None
    ) -> tuple[bool, Confidence]:
        """
        Determine if the command is a packaging command.

        A packaging command usually performs multiple tasks, such as compilation and creating the artifact.
        This function filters the build tool commands that are called from the configuration files provided as input.

        Parameters
        ----------
        cmd: BuildToolCommand
            The build tool command object.
        excluded_configs: list[str] | None
            Build tool commands that are called from these configuration files are excluded.

        Returns
        -------
        tuple[bool, Confidence]
            Return True along with the inferred confidence level if the command is a build tool command.
        """
        # Check the language.
        if cmd["language"] is not self.language:
            return False, Confidence.HIGH

        builder = self.packager if self.packager else self.builder

        if not self.match_cmd_args(cmd=cmd["command"], tools=builder, args=self.build_arg):
            return False, Confidence.HIGH

        # Check if the CI workflow is a configuration for a known tool.
        if excluded_configs and os.path.basename(cmd["ci_path"]) in excluded_configs:
            return False, Confidence.HIGH

        return True, Confidence.HIGH
