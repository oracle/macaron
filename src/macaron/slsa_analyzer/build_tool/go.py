# Copyright (c) 2023 - 2024, Oracle and/or its affiliates. All rights reserved.
# Licensed under the Universal Permissive License v 1.0 as shown at https://oss.oracle.com/licenses/upl/.

"""This module contains the Go class which inherits BaseBuildTool.

This module is used to work with repositories that have Go.
"""

from macaron.config.defaults import defaults
from macaron.dependency_analyzer.dependency_resolver import DependencyAnalyzer, NoneDependencyAnalyzer
from macaron.slsa_analyzer.build_tool.base_build_tool import BaseBuildTool, file_exists
from macaron.slsa_analyzer.build_tool.language import BuildLanguage


class Go(BaseBuildTool):
    """This class contains the information of the Go build tool."""

    def __init__(self) -> None:
        super().__init__(name="go", language=BuildLanguage.GO)

    def load_defaults(self) -> None:
        """Load the default values from defaults.ini."""
        if "builder.go" in defaults:
            for item in defaults["builder.go"]:
                if hasattr(self, item):
                    setattr(self, item, defaults.get_list("builder.go", item))

        if "builder.go.ci.deploy" in defaults:
            for item in defaults["builder.go.ci.deploy"]:
                if item in self.ci_deploy_kws:
                    self.ci_deploy_kws[item] = defaults.get_list("builder.go.ci.deploy", item)

    def is_detected(self, repo_path: str) -> bool:
        """Return True if this build tool is used in the target repo.

        Parameters
        ----------
        repo_path : str
            The path to the target repo.

        Returns
        -------
        bool
            True if this build tool is detected, else False.
        """
        go_config_files = self.build_configs + self.entry_conf
        for file in go_config_files:
            if file_exists(repo_path, file):
                return True

        return False

    def prepare_config_files(self, wrapper_path: str, build_dir: str) -> bool:
        """Prepare the necessary wrapper files for running the build.

        Go doesn't require preparation, so return true.

        Parameters
        ----------
        wrapper_path : str
            The path where all necessary wrapper files are located.
        build_dir : str
            The path of the build dir. This is where all files are copied to.

        Returns
        -------
        bool
            True if succeed else False.
        """
        return True

    def get_dep_analyzer(self, repo_path: str) -> DependencyAnalyzer:
        """Create a DependencyAnalyzer for the build tool.

        Parameters
        ----------
        repo_path: str
            The path to the target repo.

        Returns
        -------
        DependencyAnalyzer
            The DependencyAnalyzer object.
        """
        # TODO: Implement this method.
        return NoneDependencyAnalyzer()
