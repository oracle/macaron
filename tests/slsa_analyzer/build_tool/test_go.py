# Copyright (c) 2023 - 2024, Oracle and/or its affiliates. All rights reserved.
# Licensed under the Universal Permissive License v 1.0 as shown at https://oss.oracle.com/licenses/upl/.

"""This module tests the Go build functions."""

from pathlib import Path

import pytest

from macaron.code_analyzer.call_graph import BaseNode
from macaron.slsa_analyzer.build_tool.base_build_tool import BuildToolCommand
from macaron.slsa_analyzer.build_tool.go import Go
from macaron.slsa_analyzer.build_tool.language import BuildLanguage
from tests.slsa_analyzer.mock_git_utils import prepare_repo_for_testing


@pytest.mark.parametrize(
    ("folder", "file"),
    [
        ("root_go_mod", "go.mod"),
        ("no_go_mod", "dummyfile.txt"),
    ],
)
def test_get_build_dirs(snapshot: list, tmp_path: Path, go_tool: Go, folder: str, file: str) -> None:
    """Test discovering build directories."""
    # Since there's issues having 2 go.mod files in the same project, we make
    # it on the fly for this test
    proj_dir = tmp_path.joinpath(folder)
    proj_dir.mkdir(parents=True)

    with open(proj_dir.joinpath(file), "w", encoding="utf-8"):
        assert list(go_tool.get_build_dirs(str(proj_dir))) == snapshot


@pytest.mark.parametrize(
    ("folder", "file", "expected_value"),
    [
        ("root_go_mod", "go.mod", True),
        ("no_go_mod", "dummyfile.txt", False),
    ],
)
def test_go_build_tool(
    go_tool: Go, macaron_path: str, tmp_path: Path, folder: str, file: str, expected_value: bool
) -> None:
    """Test the Go build tool."""
    base_dir = Path(__file__).parent

    # Making directories with a go.mod but no actual Go project seems to cause issues
    # for the pre-commit hooks (errors like go: warning: "./..." matched no packages);
    # as such it is easiest for this test to just create/delete the go.mod files
    # macaron looks for on the fly instead of managing a proper project within the mock repos.
    proj_dir = tmp_path.joinpath(folder)
    proj_dir.mkdir(parents=True)

    with open(proj_dir.joinpath(file), "w", encoding="utf-8"):
        ctx = prepare_repo_for_testing(proj_dir, macaron_path, base_dir)
        assert go_tool.is_detected(ctx.component.repository.fs_path) == expected_value


@pytest.mark.parametrize(
    (
        "command",
        "language",
        "language_versions",
        "language_distributions",
        "ci_path",
        "reachable_secrets",
        "events",
        "excluded_configs",
        "expected_result",
    ),
    [
        (
            ["goreleaser", "release"],
            BuildLanguage.GO,
            None,
            None,
            ".github/workflows/release.yaml",
            [{"key", "pass"}],
            ["release"],
            ["codeql-analysis.yaml"],
            True,
        ),
        (
            ["goreleaser", "release"],
            BuildLanguage.GO,
            None,
            None,
            ".github/workflows/go.yaml",
            [{"key", "pass"}],
            ["push"],
            ["go.yaml"],
            False,
        ),
        (
            ["goreleaser", "release"],
            BuildLanguage.JAVASCRIPT,
            None,
            None,
            ".github/workflows/release.yaml",
            [{"key", "pass"}],
            ["push"],
            None,
            False,
        ),
    ],
)
def test_is_go_deploy_command(
    go_tool: Go,
    command: list[str],
    language: str,
    language_versions: list[str],
    language_distributions: list[str],
    ci_path: str,
    reachable_secrets: list[str],
    events: list[str],
    excluded_configs: list[str] | None,
    expected_result: bool,
) -> None:
    """Test the deploy commend detection function."""
    result, _ = go_tool.is_deploy_command(
        BuildToolCommand(
            command=command,
            language=language,
            language_versions=language_versions,
            language_distributions=language_distributions,
            language_url=None,
            ci_path=ci_path,
            step_node=BaseNode(),
            reachable_secrets=reachable_secrets,
            events=events,
        ),
        excluded_configs=excluded_configs,
    )
    assert result == expected_result


@pytest.mark.parametrize(
    (
        "command",
        "language",
        "language_versions",
        "language_distributions",
        "ci_path",
        "reachable_secrets",
        "events",
        "excluded_configs",
        "expected_result",
    ),
    [
        (
            ["go", "build"],
            BuildLanguage.GO,
            None,
            None,
            ".github/workflows/release.yaml",
            [{"key", "pass"}],
            ["release"],
            ["codeql-analysis.yaml"],
            True,
        ),
        (
            ["go", "install"],
            BuildLanguage.GO,
            None,
            None,
            ".github/workflows/release.yaml",
            [{"key", "pass"}],
            ["push"],
            ["go.yaml"],
            True,
        ),
        (
            ["go", "version"],
            BuildLanguage.GO,
            None,
            None,
            ".github/workflows/release.yaml",
            [{"key", "pass"}],
            ["release"],
            ["codeql-analysis.yaml"],
            False,
        ),
        (
            ["go", "build"],
            BuildLanguage.JAVA,
            None,
            None,
            ".github/workflows/release.yaml",
            [{"key", "pass"}],
            ["push"],
            None,
            False,
        ),
    ],
)
def test_is_go_package_command(
    go_tool: Go,
    command: list[str],
    language: str,
    language_versions: list[str],
    language_distributions: list[str],
    ci_path: str,
    reachable_secrets: list[str],
    events: list[str],
    excluded_configs: list[str] | None,
    expected_result: bool,
) -> None:
    """Test the packaging command detection function."""
    result, _ = go_tool.is_package_command(
        BuildToolCommand(
            command=command,
            language=language,
            language_versions=language_versions,
            language_distributions=language_distributions,
            language_url=None,
            ci_path=ci_path,
            step_node=BaseNode(),
            reachable_secrets=reachable_secrets,
            events=events,
        ),
        excluded_configs=excluded_configs,
    )
    assert result == expected_result
