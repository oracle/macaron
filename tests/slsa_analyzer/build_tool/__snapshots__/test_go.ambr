# serializer version: 1
# name: test_get_build_dirs[root_go_mod-go.mod]
  list([
    PosixPath('.'),
  ])
# ---
# name: test_get_build_dirs[no_go_mod-dummyfile.txt]
  list([
  ])
# ---
